toolchain/lowering/lowering_context.cpp

```diff
@@ -0,0 +1,65 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/lowering/lowering_context.h"
+
+#include "toolchain/semantics/semantics_ir.h"
+
+namespace Carbon {
+
+LoweringContext::LoweringContext(llvm::LLVMContext& llvm_context,
+                                 llvm::StringRef module_name,
+                                 const SemanticsIR& semantics_ir)
+    : llvm_context_(&llvm_context),
+      llvm_module_(std::make_unique<llvm::Module>(module_name, llvm_context)),
+      builder_(llvm_context),
+      semantics_ir_(&semantics_ir),
+      lowered_nodes_(semantics_ir_->nodes_size(), nullptr) {
+  CARBON_CHECK(!semantics_ir.has_errors())
+      << "Generating LLVM IR from invalid SemanticsIR is unsupported.";
+}
+
+auto LoweringContext::Run() -> std::unique_ptr<llvm::Module> {
+  CARBON_CHECK(llvm_module_) << "Run can only be called once.";
+
+  LowerBlock(semantics_ir_->top_node_block_id());
+
+  while (!todo_blocks_.empty()) {
+    auto [llvm_block, block_id] = todo_blocks_.pop_back_val();
+    builder_.SetInsertPoint(llvm_block);
+    LowerBlock(block_id);
+  }
+
+  return std::move(llvm_module_);
+}
```

Comment:
With this function and `LowerBlock`, this doesn't really feel like a pure context object...

I like the factoring, but maybe leave this as just `Lowering` rather than adding the context suffix?

Or maybe inline this code into LowerToLLVM?

Not a blocking comment either way. The bulk of the change to have handlers factored out and easily split further makes lots of sense.

Reply:
I'd like to handle this separate from the refactoring. I agree there's a bit of a schism, and partly I'm avoiding addressing it because I'm still not convinced about whether it's the right flow for lowering. But I know I should.

Reply:
(to be clear, not sure the block lowering approach is right, but I need to sort it out -- partly doing more lowering stuff to test theories)